virtctl, guestfs: Get rid of globals

oshoval opened the pull request:
### What this PR does
Remove globals, and some unused variables.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

oshoval commented:
/cc @0xFelix 

can you please take a look ?

oshoval commented:
Hi @brianmcarey 
the flake here returns lately during cluster-up, fyi please

https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/13020/pull-kubevirt-e2e-k8s-1.31-sig-network/1843292337933914112

brianmcarey commented:
> Hi @brianmcarey the flake here returns lately during cluster-up, fyi please
> 
> https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/13020/pull-kubevirt-e2e-k8s-1.31-sig-network/1843292337933914112

Yeah its a new flake - looking into it.

oshoval commented:
/test pull-kubevirt-e2e-k8s-1.31-sig-operator

0xFelix commented on pkg/virtctl/guestfs/guestfs.go:
```diff
@@ -162,17 +161,17 @@ func init() {
 
-func (c *guestfsCommand) run(cmd *cobra.Command, args []string) error {
-	pvc = args[0]
+func (c *guestfsCommand) run(args []string) error {
```
If you revert this change and ignore the `*cobra.Command` with `_` you can pass `c.run` directly to `RunE` in L76 without the wrapping lambda func.

0xFelix commented on pkg/virtctl/guestfs/guestfs.go:
```diff
@@ -45,47 +45,46 @@ const (
+	podParams    libguestfsPodParams
+}
+
+type libguestfsPodParams struct {
```
Not sure about this level of nesting. IMO it is fine to keep these values in the `guestfsCommand` struct.

oshoval replied:
but then need to send lot of parameters to the function, i dont think it is good idea

0xFelix commented on pkg/virtctl/guestfs/guestfs.go:
```diff
@@ -379,7 +377,7 @@ func (client *K8sClient) getPodsForPVC(pvcName, ns string) ([]corev1.Pod, error)
 }
 
-func setFSGroupLibguestfs() (*int64, error) {
+func setFSGroupLibguestfs(root bool, fsGroup string) (*int64, error) {
```
Have you considered changing this and other funcs to be receiving funcs of `guestfsCommand` instead of passing args down the chain?

oshoval replied:
i did ye, wasnt sure if it fits
i can try

0xFelix replied:
I'd still change this and the other set funcs to be receiving funcs of `guestfsCommand`.

oshoval commented:
rebased only

oshoval commented:
addressed comments, 
converted part of the functions to struct methods,
but kept some which better to just get the params they need (i.e `setFSGroupLibguestfs` and the 2 others like it)

oshoval commented:
pull-kubevirt-unit-test-s390x flakes
not related
/test pull-kubevirt-unit-test-s390x

oshoval commented:
a follow-up will remove init
this is the desired method to do so right ?
https://github.com/kubevirt/kubevirt/commit/ea87c2a20858e822a4af320d9d8d62dac0eeabf9

0xFelix commented on pkg/virtctl/guestfs/guestfs.go:
```diff
@@ -182,29 +175,30 @@ func (c *guestfsCommand) run(cmd *cobra.Command, args []string) error {
 		return err
 	}
-	defer client.removePod(namespace)
-	return client.createInteractivePodWithPVC(pvc, image, namespace, "/entrypoint.sh", []string{}, isBlock)
```
In this case I'd stick to passing `pvc` and `image` instead of the full command.

0xFelix commented on pkg/virtctl/guestfs/guestfs.go:
```diff
@@ -614,22 +605,26 @@ func createAttacher(client *K8sClient, p *corev1.Pod, command string) error {
 
-func (client *K8sClient) createInteractivePodWithPVC(pvc, image, ns, command string, args []string, isblock bool) error {
-	pod, err := createLibguestfsPod(pvc, image, command, args, kvm, isblock)
+func (client *K8sClient) createInteractivePodWithPVC(c *guestfsCommand, ns, command string, args []string, isblock bool) error {
```
I see why you started passing `c *guestfsCommand` now. IMO placing these methods on the client is a bit awkward. What do you think about change the receiver to `c *guestfsCommand` and adding a func argument for the client instead?

oshoval commented:
addressed comments

0xFelix commented on pkg/virtctl/guestfs/guestfs.go:
```diff
@@ -614,22 +605,26 @@ func createAttacher(client *K8sClient, p *corev1.Pod, command string) error {
 
-func (client *K8sClient) createInteractivePodWithPVC(pvc, image, ns, command string, args []string, isblock bool) error {
-	pod, err := createLibguestfsPod(pvc, image, command, args, kvm, isblock)
+func (c *guestfsCommand) createInteractivePodWithPVC(client *K8sClient, ns, command string, args []string, isblock bool) error {
```
Maybe in a follow up: This method's signature could be simplified a lot since it is used only in a single place and it gets passed a constant and an empty slice.

0xFelix commented on pkg/virtctl/guestfs/guestfs.go:
```diff
@@ -397,7 +389,7 @@ func setFSGroupLibguestfs() (*int64, error) {
 
 // setUIDLibugestfs returns the guestfs uid
-func setUIDLibugestfs() (*int64, error) {
+func setUIDLibugestfs(root bool, uid string) (*int64, error) {
```
Typo in func name

oshoval commented:
addressed comments

0xFelix commented on pkg/virtctl/guestfs/guestfs_test.go:
```diff
@@ -192,20 +192,17 @@ var _ = Describe("Guestfs shell", func() {
 			guestfs.SetImageInfoGetFunc(fakeGetImageInfoNoCustomURL)
-			guestfs.ImageSetFunc(kubevirtClient)
-			Expect(*guestfs.ImagePtr).To(Equal("someregistry.io/kubevirt/some-prefix-libguestfs-tools@89af657d3c226ac3083a0986e19efe70c9ccd7e7278137e9df24b9b430182aa7"))
+			Expect(guestfs.ImageSetFunc(kubevirtClient)).Should(HaveValue(Equal("someregistry.io/kubevirt/some-prefix-libguestfs-tools@89af657d3c226ac3083a0986e19efe70c9ccd7e7278137e9df24b9b430182aa7")))
```
Is `HaveEqual` really required?

oshoval replied:
yes and no

yes because theoretically the return value of ImageSetFunc can be a nil ptr
so HaveValue checks this as well

on the UT we always use fake functions that do return a value, so in this specific case Equal can be enough

imho it is nicer and more robust to have it like this, but your call

0xFelix reviewed (commented):
/approve

Being honest, I don't like the pattern with getters/setters for `ImageSetFunc` very much. IMO simple vars suffice, but this is a clean up for another PR.

oshoval commented:
> Being honest, I don't like the pattern with getters/setters for `ImageSetFunc` very much. IMO simple vars suffice, but this is a clean up for another PR.

There is already a PR for it once this PR is merged
https://github.com/kubevirt/kubevirt/pull/13133

jcanocan commented:
Nothing to add. Nice job
/lgtm
